keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2043 @@
+
+The following diagram shows how resource allocation works for a resource that
+gets defined inline in a Pod. For a full definition of ResourceClass,
+ResourceClaim and ResourceClaimTemplate see the [API](#API) section below.
+
+Several of these operations may fail without changing the system state. They
+then must be retried until they succeed or something else changes in the
+system, like for example deleting objects. These additional state transitions
+are not shown for the sake of simplicity.
+
+![allocation](./allocation.png)
+
+The flow is similar for a ResourceClaim that gets created as a stand-alone
+object by the user. In that case, the Pod reference that ResourceClaim by
+name. The ResourceClaim does not get deleted at the end and can be reused by
+another Pod and/or used by multiple different Pods at the same time (if
+supported by the driver). The resource remains allocated as long as the
+ResourceClaim doesn't get deleted.
+
+### API
+
+ResourceClaim and ResourceClass are new built-in types in a new
+`cdi.k8s.io/v1alpha1` API group. This was chosen instead of using CRDs because
+core Kubernetes components must interact with them and installation of CRDs as
+part of cluster creation is an unsolved problem.
+
+The PodSpec gets extended. Types and structs referenced from PodSpec
+(ResourceClaimTemplate, ResourceClaimSpec) must be placed in `core.k8s.io/v1`
+to avoid a cyclic dependency because `cdi.k8s.io/v1alpha1` depends on
+NodeSelector from the core API.
+
+Secrets are not part of this API: if a resource driver needs secrets, for
+example to access its own backplane, then it can define custom parameters for
+those secrets and retrieve them directly from the apiserver. This works because
+drivers are expected to be written for Kubernetes.
+
+```
+// ResourceClass is used by administrators to influence how resources
+// are allocated.
+type ResourceClass struct {
+	metav1.TypeMeta
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	metav1.ObjectMeta
+
+	// DriverName determines which resource driver is to be used for
+	// allocation of a ResourceClaim that uses this class.
+	//
+	// Resource drivers have a unique name in normal domain order
+	// (acme.example.com).
+	DriverName string
+
+	// Parameters references an arbitrary separate object that may hold
+	// parameters that will be used by the
+	// driver when allocating a resource that uses this class. The driver
+	// will be able to distinguish between parameters stored here and and
+	// those stored in ResourceClaimSpec. These parameters here can only be
+	// set by cluster administrators.
+	//
+	// The object must be cluster-scoped, so the namespace must be empty.
+	// ApiVersion and name must be set. Kind and/or resource must be set.
+	// Usually, kind is sufficient because there will be a 1:1 mapping
+	// to resource. If UID is set, then only that specific object
+	// may be used.
+	Parameters v1.ObjectReference
```

Comment:
See previous point about defining your own type for this (which means you can elide the Namespace field, for example)

Reply:
Not sure why I missed the existing `TypedLocalObjectReference` - that is much simpler and has just the right fields. I switched to that.

`APIGroup` is optional there. It seems unlikely that we will have in-tree parameter types, but perhaps we shouldn't rule it out a-priori and thus allow it to be optional?

It doesn't make a difference in YAML. Just the code that uses the type becomes more complicated.